torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,88 +62,255 @@ def reset_table(self) -> None:
-# since it does not do any contextual analysis. This means that we might incorrectly
-# find extra mutations but this is safe as it would only be incorrect to miss
-# mutations.
-class MutationTracker(ast.NodeVisitor):
-    ALLOWED_READ_FNS = {
-        "load",
-        "max_constancy",
-        "max_contiguous",
-        "multiple_of",
-        "static_print",
-        "static_assert",
-        "device_print",
-        "device_assert",
-    }
+class Op:
+    name: str
+    args: List[Union[Param, Intermediate]]
+
+
+def parse_ttir(ttir, kwargs):
+    """
+    Given a Triton emitted TTIR text, this function lexes and parses the
+    code using a minimal grammar defined inside. During the lexing/parsing,
+    we drop any constant value and type information as they are not
+    necessary to us.
+    Being able to choose what we need makes this not a general purpose TTIR
+    parser which further makes parsing much simpler.
+    """
+    # TODO(oulgen):
+    # - Support multiple functions
+    # - Support parsing of conditionals
+    # - Support parsing for/while loops
+    # - Support alternative syntax parsing similar to "tt.name"(%0, %1)
+    # - Support ops with multiple return value (e.g. %4:2 = "tt.reduce")
+
+    if ttir.count("tt.func") != 1:
+        log.debug("Multiple functions in TTIR")
+        return None
 
-    def __init__(self, infos) -> None:
-        super().__init__()
-        self.infos = infos
-        self.read_depth = 0
-        self.in_store = False
-
-    def visit_Name(self, node):
-        if node.id not in self.infos:
-            return
-        if self.read_depth:
-            pass
-        elif self.in_store:
-            self.infos[node.id].mutated = True
-        else:
-            self.infos[node.id].used_in_unknown = True
-
-    def visit_Call(self, node):
-        # TODO(oulgen): Here we assume that there exists a line called
-        # from triton import language as tl. This needs to be checked
-        # as if someones imports xyz as tl then we will incorrectly
-        # assume a mutation but this would be ok as it is only unsafe to
-        # miss a mutation.
-        if (
-            isinstance(node.func, ast.Attribute)
-            and isinstance(node.func.value, ast.Name)
-            and node.func.value.id == "tl"
+    try:
+        import lark
+        from lark import Lark, Transformer, v_args
+    except ModuleNotFoundError:
+        log.debug("Cannot parse TTIR as Lark is not installed")
+        return None
+
+    ops: Dict[Intermediate, Op] = {}
+    next_fake_intermediate = 0
+
+    # Ops looks like one of the following forms:
+    #
+    # %14 = tt.addptr %13, %4 : tensor<4x!tt.ptr<f32, 1>>, tensor<4xi32>
+    # tt.store %14, %12, %5 {cache = 1 : i32, evict = 1 : i32} : tensor<4xf32>
+    # %15 = "tt.atomic_rmw"(%14, %12, %5) <{atomic_rmw_op = 5 : i32, scope = 1 : i32, sem = 4 : i32}> : (tensor<4x!tt.ptr<f32, 1>>, tensor<4xf32>, tensor<4xi1>) -> tensor<4xf32>  # noqa: B950
+    grammar = """
```

Comment:
How do we convince ourselves this grammar is complete?  How we we detect changes in the future?

Reply:
This grammar is in fact not complete as I mentioned on the TODO at the top of the file. I'll add more rules as we need. I'm planning to work on control flow next. If the grammar changes, the tests I have added will start failing. If we fail to parse the file, we just skip on this optimization so it is always safe.

Reply:
I think, if we systematically fallback to the conservative mutation tracking if the analysis / grammar fails, we should be good?